### PR TITLE
[HybridParallel] Add ParallelCrossEntropy for TensorParallel

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

New features
### PR changes

APIs
### Describe

[HybridParallel] Add ParallelCrossEntropy for TensorParallel

### CODE REVIEW

File: python/paddle/distributed/collective.py [hunk 1]
```diff
@@ -954,6 +954,35 @@ def extra_repr(self):
...
+    if input_dims - 1 == label_dims:
+        label = paddle.unsqueeze(label, axis=-1)
+
+    if in_dygraph_mode():
+        softmax, loss = core.ops.c_softmax_with_cross_entropy(
+            logits, label, 'ring_id', ring_id, 'rank', rank, 'nranks', nranks)
+        if not return_softmax:
+            return loss
+        else:
+            return loss, softmax
```
Comment (sandyhouse) [inline python/paddle/distributed/collective.py:983]: Given an error if in static mode?

File: paddle/fluid/operators/collective/c_softmax_with_cross_entropy_op.cu [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+      MaskLabelByIndex<T, int64_t><<<blocks, threads, 0, dev_ctx.stream()>>>(
+          predicted_logits.data<T>(), softmax_2d.data<T>(),
+          labels->data<int64_t>(), start_index, end_index, N, D, nranks);
+    }
+
+    void* predict_logits_buff = predicted_logits.mutable_data<T>(place);
+    PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::ncclAllReduce(
+        predict_logits_buff, predict_logits_buff, predicted_logits.numel(),
+        platform::ToNCCLDataType(predicted_logits.type()), ncclSum,
+        comm->comm(), stream));
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/c_softmax_with_cross_entropy_op.cu:166]: should use #ifdef PADDLE_WITH_NCCL to use nccl function?

File: paddle/fluid/operators/collective/c_softmax_with_cross_entropy_op.cu [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    void* sum_exp_logits_buff = sum_exp_logits.mutable_data<T>(place);
+
+    auto eigen_sum_exp_logits = math::EigenMatrix<T>::From(sum_exp_logits);
+    eigen_sum_exp_logits.device(*dev_ctx.eigen_device()) =
+        eigen_softmax.sum(along_axis);
+
+    PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::ncclAllReduce(
+        sum_exp_logits_buff, sum_exp_logits_buff, sum_exp_logits.numel(),
+        platform::ToNCCLDataType(sum_exp_logits.type()), ncclSum, comm->comm(),
+        stream));
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/c_softmax_with_cross_entropy_op.cu:184]: As above.

**Review by sandyhouse — APPROVED** [review]

**Review by zhhsplendid — APPROVED** [review]
LGTM for ShareDataWith because the usage is similar to old softmax_with_cross_entropy_op

**Review by chenwhql — APPROVED** [review]
LGTM for PADDLE_ENFORCE in kernel

**Pull request merged**